notebooks/icrp107_dataset/icrp107_dataset.ipynb

```diff
@@ -400,25 +423,40 @@
     "for i in range(0, num_nuclides):\n",
     "    line = ffline.read(file_NDX.readline())\n",
     "    line = [i.strip() if isinstance(i,str) else i for i in line]\n",
-    "    add = {\"Radionuclide\": line[0]}\n",
+    "    name = line[0]\n",
+    "    add = {\"Radionuclide\": name}\n",
     "    add[\"Element\"] = add[\"Radionuclide\"].split(\"-\")[0]\n",
     "    add[\"Z\"], add[\"A\"] = get_Z_A(add[\"Radionuclide\"])\n",
-    "    if add[\"Radionuclide\"][-1].isalpha():\n",
-    "        add[\"Metastable_state\"] = add[\"Radionuclide\"][-1]\n",
-    "    if (line[0].endswith(\"m\") == True\n",
-    "        or line[0].endswith(\"n\") == True):\n",
-    "        add[\"Atomic_mass\"] = (line[28] +\n",
-    "                              nubase_df.loc[add[\"Radionuclide\"]].Isomer_excitation_energy)\n",
-    "    else:\n",
-    "        add[\"Atomic_mass\"] = line[28]\n",
+    "    ame_mass = ame2020_df.loc[name.strip(\"nm\")][\"Mass\"]\n",
     "    add[\"Half_life\"] = line[1]\n",
     "    unit = line[2]\n",
     "    if unit == \"us\":\n",
     "        unit = \"μs\"\n",
     "    add[\"Half_life_units\"] = unit\n",
     "    add[\"Half_life_s\"] = convert_half_life(line[1], unit)\n",
     "    add[\"Half_life_readable\"] = readable_half_life(line[1], unit)\n",
-    "\n",
+    "    if add[\"Radionuclide\"][-1].isalpha():\n",
+    "        add[\"Metastable_state\"] = add[\"Radionuclide\"][-1]\n",
+    "        # loop to check for different labelling of metastable states in NuBase as opposed to ICRP-107\n",
+    "        for meta in [\"m\",\"n\",\"p\",\"q\",\"r\"]:\n",
+    "            meta_name = name[:-1] + meta\n",
+    "            if meta_name in nubase_df.index:\n",
+    "                nubase_info = nubase_df.loc[meta_name]\n",
+    "                # account for slight differences in ICRP-107 and NuBase half-life data\n",
+    "                if isclose(add[\"Half_life_s\"], nubase_info[\"Half_life_s\"], rel_tol=.15) == True:\n",
+    "                    isomer_excitation_energy = nubase_info.Isomer_excitation_energy\n",
+    "                    break\n",
+    "                elif meta_name.endswith(\"r\"):\n",
+    "                    print(meta_name, add[\"Half_life_s\"], nubase_info[\"Half_life_s\"])\n",
+    "            if meta == \"r\":\n",
+    "                if any([issue_names in name for issue_names in [\"Ta-178m\",\"Y-84m\",\"Pm-154m\",\"Pr-134m\"]]):\n",
+    "                    # possible error in NuBase - Ta-178 and Ta-178m have swapped half-lives, same for others\n",
```

Comment:
Ahh this is another issue, isn't it.

Looking at the data, the most probable explanation is not an error in NuBase2020 or ICRP-07, but that updated nuclear data knowledge now says what was previously thought to be the metastable isomer is actually the ground state, and vice versa.

I think your solution here has the issue that it does not assign the right mass for these eight (seven*) cases. The labels are switched in Nubase, so the "Ta-178" in ICRP-07 should be taking the Nubase "Ta-178" base mass plus the Nubase "Ta-178m" excess mass. The "Ta-178m" in ICRP-07 should only be taking the Nubase "Ta-178" base mass.

This is not a major issue, as these are all obscure radionuclides and the mass differences are very small. So I'm ok without a correction.

But this follows nicely into another comment I had on this code block. Your solution loops through and finds the first Nubase2020 nuclide within 15% half-life tolerance. However I noticed there are some cases where the relative half-life difference between two states of the same nucleus is less than 15%, e.g. Au-198 & Au-198m. A more robust solution would perhaps pick out a sub-dataframe containing the ground state and all isomers of the pertinent nuclide, adding the half-life difference (abs or relative) to ICRP-07, sorting the DF by this, then picking out the first row with the closest half-life state to match with ICRP-07. This would be a fun data science exercise as Pandas has very powerful features for doing this - in fewer lines of code than you'd think!

Anyway, I'll leave the choice with you. Happy to help with this on Slack/Zoom or let it be as is!

*only 7 cases as Y-84 is not in ICRP-07.

Reply:
Thanks for the advice @alexmalins! I'll commit my attempt to fix the problem, I played with a vectorization of cycling through the matching subframes of each specific nuclide, but I am new to pandas optimization so please let me know if there are faster methods! Also, I was unable to find the licenses for NuBase and AME, please let me know if you've had any luck with locating those.